bindings/apns/create_operation.go

```diff
@@ -0,0 +1,148 @@
+package apns
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"net/http"
+
+	"github.com/dapr/components-contrib/bindings"
+	"github.com/dapr/dapr/pkg/logger"
+)
+
+type createOperation struct {
+	authorizationBuilder *authorizationBuilder
+	client               *http.Client
+	logger               logger.Logger
+	req                  *bindings.InvokeRequest
+	res                  *bindings.InvokeResponse
+	err                  error
+
+	urlPrefix           string
+	deviceToken         string
+	httpRequest         *http.Request
+	httpResponse        *http.Response
+	authorizationHeader string
+}
+
+func (op *createOperation) run() (*bindings.InvokeResponse, error) {
+	op.getDeviceToken()
```

Comment:
What's the benefit of splitting the code to granular methods ?

Reply:
One benefit I see to breaking down to multiple methods is easier testability, even if we don't unit test all of them.
As long as each function is doing it's own dedicated thing, I think this is fine.

Reply:
I think that pipeline pattern approach is good to handle the data stream or enrich the complicated data. I also used to use this pattern for processing and enriching the data. 

But this PR uses three different methods only to construct http request and save the values into each member variable. We need to maintain member variables in createOperation struct. I am not sure that it is good for maintainability in this use-case. For example, createOperation has `err` which is global error variable in this struct.

There are three parts in this code:
1. construct http header
2. send push notification
3. wraps response

I do not like to have everything in the single method, either :) It seems to have too granular methods. current error handling is not easy to understand the flow of the code.

Reply:
The `createOperation.err` variable is used to capture an error that occurs in the pipeline so that I can skip the rest of the steps and properly handle the error at the end and return it. I originally wrote it as a single method inside of `apps.go`, but it just seemed big and bulky and this approach seemed more readable. I thought it might be more acceptable for the PR review for my first contribution. I'm ok with condensing it back into a single function.

How do you want me to proceed?

Reply:
I definitely agree to have createOperation struct and `run()` method which is good pattern for the other components. But, IMO, having pipeline pattern, granular methods, and member variables definitions in struct would not be easy to understand the code and maintain it. 

We can remove the members and run can help the control of the flow, which would be easier to understand code and maintain. 

```
        req                  *bindings.InvokeRequest
	res                  *bindings.InvokeResponse
	err                  error

	httpRequest         *http.Request
	httpResponse        *http.Response
```

then run() can call the below methods... 

1. makeHTTPRequest
2. sendPushNotification
3. create response based on `sendPushNotification`

@mfcollins3 what do you think?

Reply:
I am adding the example... The below example would be simpler than using pipeline pattern. If we need to do the series of data processing, then pipeline pattern would be great. but this case is single http request call, so I recommend this way. 

```
type createOperation struct {
	authorizationBuilder *authorizationBuilder
	client               *http.Client
	logger               logger.Logger
}

func (op *createOperation) run() (*bindings.InvokeResponse, error) {
   req, err := op.makeHTTPRequest(...)
   body, err := op.sendPushNotification(req)
   if err != nil {
      return nil, err
   } else {
      // wraps result as bindings.invokeresponse
      return response, nil
   }
}

...

```

Reply:
I removed the pipeline code and merged it back into the main output binding type. Please take a look and let me know if you have any further thoughts. Thanks!